library/core/src/random/deterministic.rs

```diff
@@ -117,6 +117,9 @@ pub struct DeterministicRandomSource {
     // block counter, resulting in a 128-bit counter that will realistically
     // never roll over.
     counter_nonce: u128,
+    block: [u8; 64],
+    // The amount of bytes in block that were already used.
+    used: u8,
```

Comment:
While `u8` has sufficient range, it leads to a lot of casts in `fill_bytes` and `usize` is arguably more idiomatic for anything used for indexing/slicing. A potential reason to prefer `u8` would be a smaller memory footprint, but in this case it doesn't make a difference because of the alignment of `u128` (platform-specific but generally >= that of `usize`) and the nice round sizes of the other fields conspire to turn the "saved" bytes into padding.

Reply:
Theoretically, operations on `u8`s and comparisons between them should be a bit faster.  It's a microoptimization, of course, and one would have to test for whether it matters at all in this case.

Reply:
Most 32- and 64-bit architectures don't have native 8-bit registers or operations on them. Even when they do (x86's vestigial `al`, `bl`, `cl`, etc.), they're never any *cheaper* in any way that would be applicable here (e.g., if you operate on a lot of values at once, smaller types are nice for SIMD, but here we only have one counter). The best you can hope for is that the sign/zero extensions and truncations can be mostly optimized out and don't cost anything extra on loads/stores, which is often but not always the case.

The only exception are some odd "8 bit ALU with 16 bit address space" architectures of which only AVR has any semblance of Rust support.

Reply:
A very simple example for a bunch of targets: https://rust.godbolt.org/z/fYTqhqboq

Reply:
Indeed, the word "a bit" there was load-bearing to mean what you said:

> The best you can hope for is that the sign/zero extensions and truncations can be mostly optimized out and don't cost anything extra on loads/stores, which is often but not always the case.

Reply:
I don't follow. I'm saying that sticking with `u8` here is *never* "a bit faster" (on any target worth worrying about), it's either a bit slower or performs exactly the same.

Reply:
Yes, OK.  I misread.  I see what you mean now.  I am usually thinking about packing things for SIMD, but as you say, that's not applicable here either.  Agreed this should be `usize`.